c/make-derivation.nix

```diff
@@ -0,0 +1,172 @@
+, doxygenOutputDir ? "doc"
+, mathjax
+}:
+
+attrsOrFn:
+
+let
+  doxyfile = attrs: substitute {
+    name = "Doxyfile";
+    src = ./Doxyfile;
+    replacements = [
+      "--subst-var-by"
+      "name"
+      attrs.name
+      "--subst-var-by"
+      "version"
+      attrs.version
+      "--subst-var-by"
+      "outputDirectory"
+      attrs.doxygenOutputDir or doxygenOutputDir
+    ];
+  };
+
+
+  wrappedDoxygen = outputDir: doxyfiles: writeShellScriptBin "doxygen" ''
+    set -euo pipefail
+    doxyfiles=(${builtins.concatStringsSep " " doxyfiles})
+    : ''${componentDir=$PWD}
+    if [ -e "$componentDir/Doxyfile" ]; then
+      doxyfiles+=("$componentDir/Doxyfile")
+    fi
+    if [[ $@ =~ "--print-generated-config" ]]; then
+      ${buildPackages.bat}/bin/bat "''${doxyfiles[@]}"
+      exit 0
+    fi
+
+    # copy mathjax
+    mkdir -p "${outputDir}"/html/mathjax/
+    cp --no-preserve=mode "${mathjax}" "${outputDir}"/html/mathjax/MathJax.js
+
+    if [ $# -eq 0 ]; then
+      ${buildPackages.doxygen}/bin/doxygen <(cat "''${doxyfiles[@]}")
+    else
+      ${buildPackages.doxygen}/bin/doxygen "$@"
+    fi
+  '';
+
+  attrsOrFn' =
+    if builtins.isPath attrsOrFn then
+      import attrsOrFn
+    else
+      attrsOrFn;
+
+  attrsFn =
+    if builtins.isFunction attrsOrFn' then
+      attrsOrFn'
+    else
+      {}: attrsOrFn';
+
+  fn = args:
+    let
+      attrs = attrsFn args;
+      resolveInput = inputName: base.resolveInputs attrs.name inputName [ targetName "c" "_default" ] attrs.${inputName} or [ ];
+    in
+    (base.mkDerivation
+      ({
+        inherit stdenv;
+        doCheck = true;
+        strictDeps = true;
+        outputs = [ "out" "doc" "man" ];
+      } // attrs // (targetAttrsFn attrs) // {
+        nativeBuildInputs = [
+          buildPackages.clang-tools
+          buildPackages.valgrind
+          (wrappedDoxygen (attrs.doxygenOutputDir or doxygenOutputDir) [ (doxyfile attrs) "${(buildPackages.callPackage ./doxygen-awesome.nix {})}/Doxyfile" ])
+        ] ++ (resolveInput "nativeBuildInputs");
+
+        shellInputs = [
+          pkgs.${attrs.debugger or "gdb"}
+        ] ++ (resolveInput "shellInputs");
+
+        depsBuildBuild = resolveInput "depsBuildBuild";
+        depsBuildBuildPropagated = resolveInput "depsBuildBuildPropagated";
+        propagatedNativeBuildInputs = resolveInput "propagatedNativeBuildInputs";
+        depsBuildTarget = resolveInput "depsBuildTarget";
+        depsBuildTargetPropagated = resolveInput "depsBuildTargetPropagated";
+        depsHostHost = resolveInput "depsHostHost";
+        depsHostHostPropagated = resolveInput "depsHostHostPropagated";
+        buildInputs = resolveInput "buildInputs";
+        propagatedBuildInputs = resolveInput "propagatedBuildInputs";
+        defaultBuildInputs = resolveInput "defaultBuildInputs";
+        depsTargetTarget = resolveInput "depsTargetTarget";
+        depsTargetTargetPropagated = resolveInput "depsTargetTargetPropagated";
```

Comment:
I wonder if we could map recursively through nedryland sets and components, they are cheap to evaluate and we should more or less know where derivations are supposed to show up so maybe the lazyness of nix could make it viable. It would however make a up-front evaluation of the top level inputs to determine if they are nedryland sets/components or derivations (which might have to happen anyway?)... Not something we need to tackle now but would be interesting to test.

Reply:
Ni wait, everything is linkfarms these days so nothing can be lazy